`-DSTATIC_LINK=ON` leads to "`-lgcc_s` not found" on POSIX systems

### How to reproduce ###

1. Configure on Linux with `-DSTATIC_LINK=ON`
2. Build

### Actual behavior ###
```
[105/106] Linking CXX executable ccache
FAILED: ccache 
: && /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/GCCcore/12.3.0/bin/g++ -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -fuse-ld=bfd CMakeFiles/ccache.dir/src/main.cpp.o -o ccache  src/libccache_framework.a  -Wl,-Bdynamic  /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/zstd/1.5.5-GCCcore-12.3.0/lib/libzstd.a  -Wl,-Bstatic  -pthread  src/third_party/libthird_party.a  src/third_party/blake3/libblake3.a  -Wl,-Bdynamic  /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/hiredis/1.2.0-GCCcore-12.3.0/lib/libhiredis.a  -Wl,-Bstatic && :
/beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/binutils/2.40-GCCcore-12.3.0/bin/ld.bfd: cannot find -lgcc_s: Datei oder Verzeichnis nicht gefunden
/beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/binutils/2.40-GCCcore-12.3.0/bin/ld.bfd: /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/hiredis/1.2.0-GCCcore-12.3.0/lib/libhiredis.a(net.c.o): in function `_redisContextConnectTcp':
net.c:(.text+0x10a8): warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/binutils/2.40-GCCcore-12.3.0/bin/ld.bfd: cannot find -lgcc_s: Datei oder Verzeichnis nicht gefunden
collect2: Fehler: ld gab 1 als Ende-Status zurck
[106/106] Linking CXX executable unittest/unittest
FAILED: unittest/unittest 
: && /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/GCCcore/12.3.0/bin/g++ -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -fuse-ld=bfd unittest/CMakeFiles/unittest.dir/TestUtil.cpp.o unittest/CMakeFiles/unittest.dir/main.cpp.o unittest/CMakeFiles/unittest.dir/test_Args.cpp.o unittest/CMakeFiles/unittest.dir/test_AtomicFile.cpp.o unittest/CMakeFiles/unittest.dir/test_Config.cpp.o unittest/CMakeFiles/unittest.dir/test_Depfile.cpp.o unittest/CMakeFiles/unittest.dir/test_Hash.cpp.o unittest/CMakeFiles/unittest.dir/test_Stat.cpp.o unittest/CMakeFiles/unittest.dir/test_Util.cpp.o unittest/CMakeFiles/unittest.dir/test_argprocessing.cpp.o unittest/CMakeFiles/unittest.dir/test_ccache.cpp.o unittest/CMakeFiles/unittest.dir/test_compopt.cpp.o unittest/CMakeFiles/unittest.dir/test_compression_types.cpp.o unittest/CMakeFiles/unittest.dir/test_core_MsvcShowIncludesOutput.cpp.o unittest/CMakeFiles/unittest.dir/test_core_Statistics.cpp.o unittest/CMakeFiles/unittest.dir/test_core_StatisticsCounters.cpp.o unittest/CMakeFiles/unittest.dir/test_core_StatsLog.cpp.o unittest/CMakeFiles/unittest.dir/test_hashutil.cpp.o unittest/CMakeFiles/unittest.dir/test_storage_local_StatsFile.cpp.o unittest/CMakeFiles/unittest.dir/test_storage_local_util.cpp.o unittest/CMakeFiles/unittest.dir/test_util_BitSet.cpp.o unittest/CMakeFiles/unittest.dir/test_util_Bytes.cpp.o unittest/CMakeFiles/unittest.dir/test_util_Duration.cpp.o unittest/CMakeFiles/unittest.dir/test_util_LockFile.cpp.o unittest/CMakeFiles/unittest.dir/test_util_TextTable.cpp.o unittest/CMakeFiles/unittest.dir/test_util_TimePoint.cpp.o unittest/CMakeFiles/unittest.dir/test_util_Tokenizer.cpp.o unittest/CMakeFiles/unittest.dir/test_util_XXH3_128.cpp.o unittest/CMakeFiles/unittest.dir/test_util_XXH3_64.cpp.o unittest/CMakeFiles/unittest.dir/test_util_expected.cpp.o unittest/CMakeFiles/unittest.dir/test_util_file.cpp.o unittest/CMakeFiles/unittest.dir/test_util_path.cpp.o unittest/CMakeFiles/unittest.dir/test_util_string.cpp.o unittest/CMakeFiles/unittest.dir/test_util_zstd.cpp.o unittest/CMakeFiles/unittest.dir/test_InodeCache.cpp.o -o unittest/unittest  src/libccache_framework.a  src/third_party/libthird_party.a  src/third_party/blake3/libblake3.a  -Wl,-Bdynamic  /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/zstd/1.5.5-GCCcore-12.3.0/lib/libzstd.a  -Wl,-Bstatic  -pthread  -Wl,-Bdynamic  /beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/hiredis/1.2.0-GCCcore-12.3.0/lib/libhiredis.a  -Wl,-Bstatic && :
/beegfs/ws/1/s3248973-EasyBuild/easybuild-ml/software/binutils/2.40-GCCcore-12.3.0/bin/ld.bfd: cannot find -lgcc_s
```

### Expected behavior ###
 Build works.



### Environment ###
4.8.3


We currently (since a long time) work around this by using `'-DCMAKE_EXE_LINKER_FLAGS="-static" -DZSTD_LIBRARY="$EBROOTZSTD/lib/libzstd.a" -DHIREDIS_LIBRARY="$EBROOTHIREDIS/lib/libhiredis.a"` but would love to use the simpler `STATIC_LINK` option

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -69,6 +69,7 @@ endif()
 include(CIBuildType)
 include(DefaultBuildType)
 include(UseFastestLinker)
+include(StaticLinkSupport)
 include(StandardSettings)
 include(StandardWarnings)
 
@@ -79,35 +80,6 @@ include(StandardWarnings)
 include(InstallDirs)
 include(GenerateConfigurationFile)
 include(GenerateVersionFile)
-
-#
-# Static link configuration
-#
-
-set(STATIC_LINK_DEFAULT OFF)
-if(WIN32)
-  set(STATIC_LINK_DEFAULT ON)
-endif()
-
-option(STATIC_LINK "Prefer linking libraries statically" ${STATIC_LINK_DEFAULT})
-
-if(STATIC_LINK)
-  list(INSERT CMAKE_FIND_LIBRARY_SUFFIXES 0 "${CMAKE_STATIC_LIBRARY_SUFFIX}")
-  set(CMAKE_LINK_SEARCH_START_STATIC ON)
-
-  # Link MSVC runtime statically.
-  if(MSVC)
-    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
-  # Link MINGW runtime statically.
-  elseif(WIN32)
-    if((CMAKE_CXX_COMPILER_ID STREQUAL GNU) OR (CMAKE_CXX_COMPILER_ID STREQUAL Clang))
-      list(APPEND CCACHE_EXTRA_LIBS -static-libgcc -static-libstdc++ -static -lwinpthread -dynamic)
-    endif()
-    if(CMAKE_CXX_COMPILER_ID STREQUAL Clang)
-      list(APPEND CCACHE_EXTRA_LIBS -fuse-ld=lld)
-    endif()
-  endif()
-endif()
 
 #
 # Third party

--- a/cmake/StaticLinkSupport.cmake
+++ b/cmake/StaticLinkSupport.cmake
@@ -1,0 +1,35 @@
+# This sets up support for maximum static linking of dependent libraries on
+# supported platforms.
+
+set(STATIC_LINK_DEFAULT OFF)
+if(WIN32)
+  set(STATIC_LINK_DEFAULT ON)
+endif()
+
+option(STATIC_LINK "Link most libraries statically" ${STATIC_LINK_DEFAULT})
+
+if(NOT STATIC_LINK)
+  return()
+endif()
+
+set(CMAKE_LINK_SEARCH_START_STATIC ON)
+set(CMAKE_FIND_LIBRARY_SUFFIXES "${CMAKE_STATIC_LIBRARY_SUFFIX}")
+
+if(WIN32)
+  # Link MSVC runtime statically.
+  if(MSVC)
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+  # Link MINGW runtime statically.
+  else()
+    if(CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang)\$")
+      list(APPEND CCACHE_EXTRA_LIBS -static-libgcc -static-libstdc++ -static -lwinpthread -dynamic)
+    endif()
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+      list(APPEND CCACHE_EXTRA_LIBS -fuse-ld=lld)
+    endif()
+  endif()
+else()
+  if(CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang)\$")
+    list(APPEND CCACHE_EXTRA_LIBS -static-libgcc -static-libstdc++)
+  endif()
+endif()

